sklearn/preprocessing/_target_encoder_fast.pyx

```diff
@@ -0,0 +1,136 @@
+            sums[X_int_tmp] += y[sample_idx]
+            counts[X_int_tmp] += 1.0
+
+        current_encoding = np.empty(shape=n_cats, dtype=np.float64)
+        for cat_idx in range(n_cats):
+            current_encoding[cat_idx] = sums[cat_idx] / counts[cat_idx]
+        encodings.append(current_encoding)
+    return encodings
+
+
+def _fit_encoding_fast_auto_smooth(
+    INT_DTYPE[:, ::1] X_int,
+    Y_DTYPE[:] y,
+    cnp.int64_t[::1] n_categories,
+    double y_mean,
+    double y_variance,
+):
+    """Fit a target encoding on X_int and y with auto smoothing.
+
+    This implementation uses Eq 5 and 6 from [1].
+
+    [1]: Micci-Barreca, Daniele. "A preprocessing scheme for high-cardinality
+         categorical attributes in classification and prediction problems"
+    """
+    cdef:
+        list encodings = []
+        cnp.int64_t sample_idx, feat_idx, cat_idx, n_cats
+        INT_DTYPE X_int_tmp
+        int n_samples = X_int.shape[0]
+        int n_features = X_int.shape[1]
+        cnp.int64_t max_n_cats = np.max(n_categories)
+        double[::1] means = np.empty(max_n_cats, dtype=np.float64)
+        cnp.int64_t[::1] counts = np.empty(max_n_cats, dtype=np.int64)
+        double[::1] sum_of_squared_diffs = np.empty(max_n_cats, dtype=np.float64)
+        double lambda_
+
+    for feat_idx in range(n_features):
+        n_cats = n_categories[feat_idx]
+
+        for cat_idx in range(n_cats):
+            means[cat_idx] = 0.0
+            counts[cat_idx] = 0
+            sum_of_squared_diffs[cat_idx] = 0.0
+
+        # first pass to compute the mean
+        for sample_idx in range(n_samples):
+            X_int_tmp = X_int[sample_idx, feat_idx]
+
+            # -1 are unknown categories, which are not counted
+            if X_int_tmp == -1:
+                continue
+            counts[X_int_tmp] += 1
+            means[X_int_tmp] += y[sample_idx]
+
+        for cat_idx in range(n_cats):
+            means[cat_idx] /= counts[cat_idx]
+
+        # second pass to compute the sum of squared differences
+        for sample_idx in range(n_samples):
+            X_int_tmp = X_int[sample_idx, feat_idx]
+            if X_int_tmp == -1:
+                continue
+            sum_of_squared_diffs[X_int_tmp] += pow(y[sample_idx] - means[X_int_tmp], 2.0)
+
+        current_encoding = np.empty(shape=n_cats, dtype=np.float64)
+        for cat_idx in range(n_cats):
+            lambda_ = (
+                y_variance * counts[cat_idx] /
+                (y_variance * counts[cat_idx] + sum_of_squared_diffs[cat_idx] / counts[cat_idx])
+            )
+            if isnan(lambda_):
+                # A nan can happen when:
+                # 1. counts[cat_idx] == 0
+                # 2. y_variance == 0 and sum_of_squared_diffs[cat_idx] == 0
+                current_encoding[cat_idx] = y_mean
+            else:
+                current_encoding[cat_idx] = lambda_ * means[cat_idx] + (1 - lambda_) * y_mean
+        encodings.append(current_encoding)
```

Comment:
The overall data structure is a "jagged array" because each categorical feature may have different cardinalities. I updated this PR and kept the Python list to store the ndarrays and a `double**` to have access to the pre-allocated encodings without the gil.

(I considered using C++ vector instead of `double**`, but this seems simple enough to stick with `malloc/free` and not require a C++ compiler)

Reply:
I see you changed your mind and decided to use a vector instead in 665e2e9. I suppose this makes the memory management cleaner as I assume that Cython will do the ref counting automatically for the returned datastructure.

Reply:
In both implementations, Python is handling the reference counting of the output. For C++, there is a temporary `vector[double *]` that points to the outputs. For C, there was a `double**` that points to the outputs. I do not think setting pointers to the output data would increment the Python reference count, because pointers are not PyObjects.

The biggest change with C++ is that `vector[double*]` is [RAII](https://en.cppreference.com/w/cpp/language/raii) and has it's own C++ reference count and freed when it gets out of scope.